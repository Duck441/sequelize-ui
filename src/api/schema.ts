import { Schema } from '@src/core/schema'
import { versionedName } from '@src/core/utils/string'
import shortid from 'shortid'

export const SCHEMA_NOT_FOUND_ERROR = '[Schema Api Error] Schema not found'

export async function listSchemas(): Promise<Schema[]> {
  return (await get<Schema[]>(schemasKey())) || []
}

export async function getSchema(id: string): Promise<Schema> {
  const schemas = await listSchemas()
  const schema = schemas.find((s) => s.id === id)
  return schema || Promise.reject(new Error(SCHEMA_NOT_FOUND_ERROR))
}

export async function createSchema(schemaPayload: Omit<Schema, 'id'>): Promise<Schema> {
  const schemas = await listSchemas()
  const name = versionedName(
    schemaPayload.name,
    schemas.map((s) => s.name),
  )
  const id = shortid()
  const schema: Schema = { ...schemaPayload, id, name }
  await set(schemasKey(), [...schemas, schema])
  return schema
}

export async function updateSchema(schema: Schema): Promise<Schema> {
  const schemas = await listSchemas()
  const existingSchema = schemas.find((s) => s.id === schema.id)

  if (!existingSchema) {
    return Promise.reject(new Error(SCHEMA_NOT_FOUND_ERROR))
  }

  const name = versionedName(
    schema.name,
    schemas.map((s) => s.name),
  )

  const updatedSchema: Schema = removeTargetingAssociations(existingSchema, { ...schema, name })
  const updatedSchemas = schemas.map((s) => (s.id === schema.id ? updatedSchema : s))

  await set(schemasKey(), updatedSchemas)
  return updatedSchema
}

function removeTargetingAssociations(schema: Schema, updatedSchema: Schema): Schema {
  const removedModelIds = schema.models
    .filter((m) => !updatedSchema.models.some((um) => um.id === m.id))
    .map((m) => m.id)

  const models = updatedSchema.models.map((m) => ({
    ...m,
    associations: m.associations.filter((a) => !removedModelIds.includes(a.targetModelId)),
  }))

  return { ...updatedSchema, models }
}

export async function deleteSchema(id: string): Promise<void> {
  const schemas = await listSchemas()
  if (schemas.length) {
    const updatedSchemas = schemas.filter((s) => s.id !== id)
    await set(schemasKey(), updatedSchemas)
  }
}

export async function clearSchemas(): Promise<void> {
  return await remove(schemasKey())
}

function get<T>(key: string): Promise<T | null> {
  return new Promise<T | null>((resolve, reject) => {
    try {
      const item: string | null = localStorage.getItem(lsKey(key))
      if (item === null) return resolve(item)
      const result: T = JSON.parse(item)
      resolve(result)
    } catch (e) {
      reject(e)
    }
  })
}

function set<T>(key: string, value: T): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    try {
      const payload = JSON.stringify(value)
      localStorage.setItem(lsKey(key), payload)
      resolve()
    } catch (e) {
      reject(e)
    }
  })
}

function remove(key: string): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    try {
      localStorage.removeItem(lsKey(key))
      resolve()
    } catch (e) {
      reject(e)
    }
  })
}

function schemasKey(): string {
  return '/schemas'
}

const NAMESPACE = `__SEQUELIZEUI__`
function lsKey(key: string): string {
  return NAMESPACE + key
}
