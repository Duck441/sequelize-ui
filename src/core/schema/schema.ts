import { now } from '@src/utils/dateTime'
import shortid from 'shortid'
import { Association, AssociationTypeType } from './association'
import { DataType, DataTypeType } from './dataType'

export type Schema = {
  id: string
  name: string
  models: Model[]
  createdAt: string
  updatedAt: string
}

export type Model = {
  id: string
  name: string
  fields: Field[]
  associations: Association[]
  createdAt: string
  updatedAt: string
}

export type Field = {
  id: string
  name: string
  type: DataType
  primaryKey?: boolean
  required?: boolean
  unique?: boolean
  generated?: boolean
}

export function emptySchema(): Schema {
  const time = now()
  return {
    id: '',
    name: '',
    models: [],
    createdAt: time,
    updatedAt: time,
  }
}

export function emptyModel(): Model {
  const time = now()
  return {
    id: shortid(),
    name: '',
    fields: [],
    associations: [],
    createdAt: time,
    updatedAt: time,
  }
}

export function emptyField(): Field {
  return { id: shortid(), name: '', type: { type: DataTypeType.String } }
}

export function emptyAssociation(
  sourceModelId: Model['id'],
  targetModelId: Model['id'],
): Association {
  return {
    id: shortid(),
    sourceModelId,
    type: { type: AssociationTypeType.BelongsTo },
    targetModelId,
  }
}
