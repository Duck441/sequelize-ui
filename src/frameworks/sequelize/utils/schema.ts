import { DbOptions } from '@src/core/database'
import { Field, Model, Schema } from '@src/core/schema'
import { idField, prefixPk } from './field'

type NormalizeSchemaArgs = {
  schema: Schema
  dbOptions: DbOptions
}
export const normalizeSchema = ({ schema, dbOptions }: NormalizeSchemaArgs): Schema => {
  return { ...schema, models: schema.models.map((model) => normalizeModel({ model, dbOptions })) }
}

type NormalizeModelArgs = {
  model: Model
  dbOptions: DbOptions
}
const normalizeModel = ({ model, dbOptions }: NormalizeModelArgs): Model => {
  return { ...model, fields: getFieldsWithPk({ model, dbOptions }) }
}

type GetFieldsWithIdArgs = {
  model: Model
  dbOptions: DbOptions
}
function getFieldsWithPk({ model, dbOptions }: GetFieldsWithIdArgs): Field[] {
  const pks = model.fields.filter((f) => f.primaryKey)

  // Don't apply any prefixes if pk is composite
  if (pks.length > 1) return model.fields
  // Add explicit pk field for TypeScript classes
  if (pks.length === 0) return [idField({ model, dbOptions }), ...model.fields]
  // Prefix pk with model name if pk is standard  format and prefixPk option is true
  return model.fields.map((field) => prefixPk({ field, model, dbOptions }))
}
