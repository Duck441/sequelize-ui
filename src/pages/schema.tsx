import {
  DatabaseCaseStyle,
  DatabaseNounForm,
  DatabaseOptions,
  SqlDialect,
} from '@src/core/database'
import DbOptionsForm from '@src/ui/components/DbOptionsForm'
import Layout from '@src/ui/components/Layout'
import ModelModule from '@src/ui/components/ModelModule'
import SchemaModule from '@src/ui/components/SchemaModule'
import useSchemaState, { SchemaEditStateType } from '@src/ui/hooks/useEditSchema'
import useGeneratedCode from '@src/ui/hooks/useGeneratedCode'
import dynamic from 'next/dynamic'
import React, { useCallback, useState } from 'react'

const CodeViewer = dynamic(() => import('@src/ui/components/CodeViewer'))

function SchemaPage(): React.ReactElement {
  return (
    <Layout title="Schema | Sequelize UI">
      <SchemaPageContent />
    </Layout>
  )
}

const defaultDbOptions: DatabaseOptions = {
  sqlDialect: SqlDialect.Postgres,
  caseStyle: DatabaseCaseStyle.Snake,
  nounForm: DatabaseNounForm.Singular,
  timestamps: true,
}

function SchemaPageContent(): React.ReactElement {
  const {
    schema,
    schemas,
    editState,
    error,
    edit,
    update,
    destroy,
    addModel,
    editModel,
    updateModel,
    deleteModel,
    cancel,
  } = useSchemaState()

  const [dbOptions, setDbOptions] = useState<DatabaseOptions>(defaultDbOptions)
  const { root } = useGeneratedCode({ schema, dbOptions })
  const [viewCode, setViewCode] = useState<boolean>(false)
  const handleClickViewCode = useCallback(() => setViewCode((x) => !x), [])

  if (error) return <p>{error}</p>
  if (schema === undefined) return <p>Loading Schemas</p>

  return (
    <>
      <button onClick={handleClickViewCode}>{viewCode ? 'Hide Code' : 'View Code'}</button>
      <SchemaModule
        schema={schema}
        schemas={schemas || []}
        editing={editState.type === SchemaEditStateType.EditingSchema}
        onEdit={edit}
        onUpdate={update}
        onCancel={cancel}
      />

      <button onClick={destroy}>Delete</button>
      {viewCode && root && (
        <>
          <DbOptionsForm dbOptions={dbOptions} onChange={setDbOptions} />
          <CodeViewer cacheKey={schema.id} root={root} />
        </>
      )}
      <h3>Models</h3>
      <button type="button" onClick={addModel}>
        Add model
      </button>
      {schema.models.length ? (
        <ul>
          {schema.models.map((model) => (
            <ModelModule
              key={`model-module-${model.id}`}
              model={model}
              schema={schema}
              editing={
                editState.type === SchemaEditStateType.EditingModel && editState.id === model.id
              }
              disabled={
                editState.type === SchemaEditStateType.EditingSchema ||
                (editState.type === SchemaEditStateType.EditingModel && editState.id !== model.id)
              }
              onRequestEdit={editModel}
              onRequestDelete={deleteModel}
              onRequestCancel={cancel}
              onChange={updateModel}
            />
          ))}
        </ul>
      ) : null}
    </>
  )
}

export default SchemaPage
