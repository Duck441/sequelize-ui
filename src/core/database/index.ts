import { pascalCase, snakeCase } from 'change-case'

export type DatabaseOptions = {
  sqlDialect: SqlDialect
  timestamps: boolean
  caseStyle: DatabaseCaseStyle
  nounForm: DatabaseNounForm
}

export enum DatabaseCaseStyle {
  Snake = 'snake',
  Camel = 'camel',
}

export enum DatabaseNounForm {
  Singular = 'singular',
  Plural = 'plural',
}

export enum SqlDialect {
  MariaDb = 'mariadb',
  MsSql = 'mssql',
  MySql = 'mysql',
  Postgres = 'postgres',
  Sqlite = 'sqlite',
}

export const displaySqlDialect = (dialect: SqlDialect): string => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'MariaDB'
    case SqlDialect.MsSql:
      return 'SQL Server'
    case SqlDialect.MySql:
      return 'MySQL'
    case SqlDialect.Postgres:
      return 'PostgreSQL'
    case SqlDialect.Sqlite:
      return 'SQLite'
  }
}

export const sqlDialectEnvVar = (dialect: SqlDialect): string => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'MARIADB'
    case SqlDialect.MsSql:
      return 'MSSQL'
    case SqlDialect.MySql:
      return 'MYSQL'
    case SqlDialect.Postgres:
      return 'POSTGRES'
    case SqlDialect.Sqlite:
      return 'SQLITE'
  }
}

export const parseSqlDialect = (dialect: string): SqlDialect | null => {
  switch (dialect) {
    case 'mariadb':
      return SqlDialect.MariaDb
    case 'mssql':
      return SqlDialect.MsSql
    case 'mysql':
      return SqlDialect.MySql
    case 'postgres':
      return SqlDialect.Postgres
    case 'sqlite':
      return SqlDialect.Sqlite
    default:
      return null
  }
}

export const defaultSqlDialectPort = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return '3306'
    case SqlDialect.MsSql:
      return '1433'
    case SqlDialect.MySql:
      return '3306'
    case SqlDialect.Postgres:
      return '5432'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectUsername = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'root'
    case SqlDialect.MsSql:
      return 'sa'
    case SqlDialect.MySql:
      return 'root'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectPassword = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'root'
    case SqlDialect.MsSql:
      return 'Password1'
    case SqlDialect.MySql:
      return 'root'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectDatabase = (name: string, dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return snakeCase(name)
    case SqlDialect.MsSql:
      return pascalCase(name)
    case SqlDialect.MySql:
      return snakeCase(name)
    case SqlDialect.Postgres:
      return snakeCase(name)
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectHost = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'localhost'
    case SqlDialect.MsSql:
      return 'localhost'
    case SqlDialect.MySql:
      return 'localhost'
    case SqlDialect.Postgres:
      return 'localhost'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectStorage = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return null
    case SqlDialect.MsSql:
      return null
    case SqlDialect.MySql:
      return null
    case SqlDialect.Postgres:
      return null
    case SqlDialect.Sqlite:
      return '.tmp/data.db'
  }
}

export const displayDatabaseNounForm = (nounForm: DatabaseNounForm): string => {
  switch (nounForm) {
    case DatabaseNounForm.Singular:
      return 'singular'
    case DatabaseNounForm.Plural:
      return 'plural'
  }
}

export const displayDatabaseCaseStyle = (caseStyle: DatabaseCaseStyle): string => {
  switch (caseStyle) {
    case DatabaseCaseStyle.Snake:
      return 'snake'
    case DatabaseCaseStyle.Camel:
      return 'camel'
  }
}

export const defaultDbOptions: DatabaseOptions = {
  sqlDialect: SqlDialect.Postgres,
  timestamps: true,
  caseStyle: DatabaseCaseStyle.Snake,
  nounForm: DatabaseNounForm.Plural,
}

export const MAX_IDENTIFIER_LENGTH = 63
