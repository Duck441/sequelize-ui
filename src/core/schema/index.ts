export * from './association'
export * from './dataType'
export * from './schema'
export * from './validation'
