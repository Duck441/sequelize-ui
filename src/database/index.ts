import { pascalCase, snakeCase } from 'change-case'

export type DatabaseOptions = {
  sqlDialect: SqlDialect
  timestamps: boolean
  caseStyle: 'snake' | 'camel'
  nounForm: 'singular' | 'plural'
}

export enum SqlDialect {
  MariaDb = 'MARIADB',
  MsSql = 'MSSQL',
  MySql = 'MYSQL',
  Postgres = 'POSTGRES',
  Sqlite = 'SQLITE',
}

export const displaySqlDialect = (dialect: SqlDialect): string => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'mariadb'
    case SqlDialect.MsSql:
      return 'mssql'
    case SqlDialect.MySql:
      return 'mysql'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return 'sqlite'
  }
}

export const defaultSqlDialectPort = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return '3306'
    case SqlDialect.MsSql:
      return '1433'
    case SqlDialect.MySql:
      return '3306'
    case SqlDialect.Postgres:
      return '5432'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectUsername = (dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return 'root'
    case SqlDialect.MsSql:
      return 'sa'
    case SqlDialect.MySql:
      return 'root'
    case SqlDialect.Postgres:
      return 'postgres'
    case SqlDialect.Sqlite:
      return null
  }
}

export const defaultSqlDialectDatabase = (name: string, dialect: SqlDialect): string | null => {
  switch (dialect) {
    case SqlDialect.MariaDb:
      return snakeCase(name)
    case SqlDialect.MsSql:
      return pascalCase(name)
    case SqlDialect.MySql:
      return snakeCase(name)
    case SqlDialect.Postgres:
      return snakeCase(name)
    case SqlDialect.Sqlite:
      return null
  }
}
