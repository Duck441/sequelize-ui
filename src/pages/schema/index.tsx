import { deleteSchema, getSchema, updateSchema } from '@src/api/schema'
import { Schema } from '@src/core/schema'
import { goTo } from '@src/routing/navigation'
import { indexRoute, parseSchemaId } from '@src/routing/routes'
import Flyout from '@src/ui/components/Flyout'
import SchemaFlyout from '@src/ui/components/SchemaFlyout'
import useAsync from '@src/ui/hooks/useAsync'
import { useRouter } from 'next/router'
import React from 'react'

export default function SchemaPage(): React.ReactElement {
  const [schema, setSchema] = React.useState<Schema>()
  const router = useRouter()
  // TODO: Navigate home with error message when no id
  const schemaId = parseSchemaId(router.query)

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const getData = React.useCallback(() => getSchema(schemaId!), [schemaId])

  const { refetch } = useAsync({ getData, skip: !schemaId, onLoad: setSchema })

  const handleChange = async (schema: Schema) => {
    const updated = await updateSchema(schema)
    setSchema(updated)
    refetch()
    return updated
  }

  const handleDelete = async () => {
    schema && (await deleteSchema(schema.id))
    goTo(indexRoute())
  }

  const handleCancel = () => goTo(indexRoute())

  return schema ? (
    <SchemaFlyout
      schema={schema}
      schemas={[]}
      onChange={handleChange}
      onDelete={handleDelete}
      onClickClose={handleCancel}
    />
  ) : (
    <Flyout title="" onClickClose={handleCancel} controls={<></>} />
  )
}
