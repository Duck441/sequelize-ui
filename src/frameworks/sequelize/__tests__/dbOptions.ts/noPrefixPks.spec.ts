import { defaultDbOptions } from '@src/core/database'
import schema from '@src/frameworks/__fixtures__/schemas/pks'
import { SequelizeFramework } from '../..'

describe('Sequelize Framework', () => {
  describe('no prefix pks options', () => {
    it('generates correct code', () => {
      const code = SequelizeFramework.generate({
        schema,
        dbOptions: {
          ...defaultDbOptions,
          prefixPks: false,
        },
      })
      expect(code).toMatchSnapshot()
    })
  })
})
