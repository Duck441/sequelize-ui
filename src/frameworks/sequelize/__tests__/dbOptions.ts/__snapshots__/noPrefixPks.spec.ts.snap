// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sequelize Framework no prefix pks options generates correct code 1`] = `
Object {
  "files": Array [
    Object {
      "files": Array [
        Object {
          "content": "module.exports = {
  development: {
    dialect: 'postgres',
    database: process.env.POSTGRES_DB_NAME || 'fields',
    username: process.env.POSTGRES_DB_USERNAME || 'postgres',
    password: process.env.POSTGRES_DB_PASSWORD || 'postgres',
    host: process.env.POSTGRES_DB_HOST || 'localhost',
    port: parseInt(process.env.POSTGRES_DB_PORT || '5432')
  },
  test: {
    dialect: 'postgres',
    database: process.env.POSTGRES_DB_NAME || 'fields',
    username: process.env.POSTGRES_DB_USERNAME || 'postgres',
    password: process.env.POSTGRES_DB_PASSWORD || 'postgres',
    host: process.env.POSTGRES_DB_HOST || 'localhost',
    port: parseInt(process.env.POSTGRES_DB_PORT || '5432')
  },
  production: {
    dialect: 'postgres',
    database: process.env.POSTGRES_DB_NAME,
    username: process.env.POSTGRES_DB_USERNAME,
    password: process.env.POSTGRES_DB_PASSWORD,
    host: process.env.POSTGRES_DB_HOST,
    port: parseInt(process.env.POSTGRES_DB_PORT)
  }
}",
          "name": "config.js",
          "type": "FILE",
        },
      ],
      "name": "config",
      "type": "DIRECTORY",
    },
    Object {
      "files": Array [
        Object {
          "content": "const DataTypes = require('sequelize').DataTypes

module.exports = {
  up: async (queryInterface, Sequelize) => {
    await queryInterface.createTable('defaults', {
      id: {
        type: DataTypes.INTEGER,
        field: 'id',
        primaryKey: true
      },
      createdAt: {
        type: DataTypes.DATE,
        field: 'created_at'
      },
      updatedAt: {
        type: DataTypes.DATE,
        field: 'updated_at'
      }
    })
  },
  down: async (queryInterface, Sequelize) => {
    await queryInterface.dropTable('defaults');
  },
};",
          "name": "20200401000000-create-defaults.js",
          "type": "FILE",
        },
        Object {
          "content": "const DataTypes = require('sequelize').DataTypes

module.exports = {
  up: async (queryInterface, Sequelize) => {
    await queryInterface.createTable('ids', {
      id: {
        type: DataTypes.INTEGER,
        field: 'id',
        primaryKey: true
      },
      createdAt: {
        type: DataTypes.DATE,
        field: 'created_at'
      },
      updatedAt: {
        type: DataTypes.DATE,
        field: 'updated_at'
      }
    })
  },
  down: async (queryInterface, Sequelize) => {
    await queryInterface.dropTable('ids');
  },
};",
          "name": "20200401000010-create-ids.js",
          "type": "FILE",
        },
        Object {
          "content": "const DataTypes = require('sequelize').DataTypes

module.exports = {
  up: async (queryInterface, Sequelize) => {
    await queryInterface.createTable('prefixeds', {
      id: {
        type: DataTypes.INTEGER,
        field: 'id',
        primaryKey: true
      },
      createdAt: {
        type: DataTypes.DATE,
        field: 'created_at'
      },
      updatedAt: {
        type: DataTypes.DATE,
        field: 'updated_at'
      }
    })
  },
  down: async (queryInterface, Sequelize) => {
    await queryInterface.dropTable('prefixeds');
  },
};",
          "name": "20200401000020-create-prefixeds.js",
          "type": "FILE",
        },
        Object {
          "content": "const DataTypes = require('sequelize').DataTypes

module.exports = {
  up: async (queryInterface, Sequelize) => {
    await queryInterface.createTable('nonstandards', {
      otherId: {
        type: DataTypes.INTEGER,
        field: 'other_id',
        primaryKey: true
      },
      createdAt: {
        type: DataTypes.DATE,
        field: 'created_at'
      },
      updatedAt: {
        type: DataTypes.DATE,
        field: 'updated_at'
      }
    })
  },
  down: async (queryInterface, Sequelize) => {
    await queryInterface.dropTable('nonstandards');
  },
};",
          "name": "20200401000030-create-nonstandards.js",
          "type": "FILE",
        },
      ],
      "name": "migrations",
      "type": "DIRECTORY",
    },
    Object {
      "files": Array [
        Object {
          "content": "import Sequelize, { DataTypes, Model, Optional } from 'sequelize'

export interface DefaultAttributes {
  
}

export type DefaultId = number
export type DefaultCreationAttributes = DefaultAttributes & { id?: number }

export class Default
  extends Model<DefaultAttributes, DefaultCreationAttributes>
  implements DefaultAttributes {
  public readonly id?: number
  
  static initModel(sequelize: Sequelize.Sequelize): typeof Default {
    Default.init({
      
    }, {
      sequelize
    })
    
    return Default
  }
}
",
          "name": "Default.ts",
          "type": "FILE",
        },
        Object {
          "content": "import Sequelize, { DataTypes, Model, Optional } from 'sequelize'

export interface IdAttributes {
  id?: number
}

export type IdPk = 'id'
export type IdId = IdAttributes[IdPk]
export type IdCreationAttributes = Optional<IdAttributes, IdPk>

export class Id
  extends Model<IdAttributes, IdCreationAttributes>
  implements IdAttributes {
  public readonly id?: number
  
  static initModel(sequelize: Sequelize.Sequelize): typeof Id {
    Id.init({
      id: {
        type: DataTypes.INTEGER,
        primaryKey: true
      }
    }, {
      sequelize
    })
    
    return Id
  }
}
",
          "name": "Id.ts",
          "type": "FILE",
        },
        Object {
          "content": "import type { Sequelize, Model } from 'sequelize'
import { Default } from './Default'
import type { DefaultAttributes, DefaultCreationAttributes } from './Default'
import { Id } from './Id'
import type { IdAttributes, IdCreationAttributes } from './Id'
import { Prefixed } from './Prefixed'
import type { PrefixedAttributes, PrefixedCreationAttributes } from './Prefixed'
import { Nonstandard } from './Nonstandard'
import type { NonstandardAttributes, NonstandardCreationAttributes } from './Nonstandard'

export {
  Default,
  Id,
  Prefixed,
  Nonstandard
}

export type {
  DefaultAttributes,
  DefaultCreationAttributes,
  IdAttributes,
  IdCreationAttributes,
  PrefixedAttributes,
  PrefixedCreationAttributes,
  NonstandardAttributes,
  NonstandardCreationAttributes
}

export function initModels(sequelize: Sequelize) {
  Default.initModel(sequelize)
  Id.initModel(sequelize)
  Prefixed.initModel(sequelize)
  Nonstandard.initModel(sequelize)

  return {
    Default,
    Id,
    Prefixed,
    Nonstandard
  }
}
",
          "name": "index.ts",
          "type": "FILE",
        },
        Object {
          "content": "import Sequelize, { DataTypes, Model, Optional } from 'sequelize'

export interface NonstandardAttributes {
  otherId?: number
}

export type NonstandardPk = 'otherId'
export type NonstandardId = NonstandardAttributes[NonstandardPk]
export type NonstandardCreationAttributes = Optional<NonstandardAttributes, NonstandardPk>

export class Nonstandard
  extends Model<NonstandardAttributes, NonstandardCreationAttributes>
  implements NonstandardAttributes {
  public readonly otherId?: number
  
  static initModel(sequelize: Sequelize.Sequelize): typeof Nonstandard {
    Nonstandard.init({
      otherId: {
        type: DataTypes.INTEGER,
        primaryKey: true
      }
    }, {
      sequelize
    })
    
    return Nonstandard
  }
}
",
          "name": "Nonstandard.ts",
          "type": "FILE",
        },
        Object {
          "content": "import Sequelize, { DataTypes, Model, Optional } from 'sequelize'

export interface PrefixedAttributes {
  id?: number
}

export type PrefixedPk = 'id'
export type PrefixedId = PrefixedAttributes[PrefixedPk]
export type PrefixedCreationAttributes = Optional<PrefixedAttributes, PrefixedPk>

export class Prefixed
  extends Model<PrefixedAttributes, PrefixedCreationAttributes>
  implements PrefixedAttributes {
  public readonly id?: number
  
  static initModel(sequelize: Sequelize.Sequelize): typeof Prefixed {
    Prefixed.init({
      id: {
        type: DataTypes.INTEGER,
        primaryKey: true
      }
    }, {
      sequelize
    })
    
    return Prefixed
  }
}
",
          "name": "Prefixed.ts",
          "type": "FILE",
        },
      ],
      "name": "models",
      "type": "DIRECTORY",
    },
    Object {
      "content": "node_modules/
dist
npm-debug.log*
yarn-debug.log*
yarn-error.log*
.npm
.tmp
",
      "name": ".gitignore",
      "type": "FILE",
    },
    Object {
      "content": "import { Sequelize, Options } from 'sequelize'
import configs from './config/config.js'

const env = process.env.NODE_ENV || 'development'
const config = (configs as {[key: string]: Options})[env]

const db: Sequelize = new Sequelize({
  ...config,
  define: {
    underscored: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at'
  }
})

export default db
",
      "name": "db.ts",
      "type": "FILE",
    },
    Object {
      "content": "{
  \\"name\\": \\"fields\\",
  \\"version\\": \\"0.0.1\\",
  \\"description\\": \\"\\",
  \\"main\\": \\"server.ts\\",
  \\"scripts\\": {
    \\"test\\": \\"echo \\\\\\"Error: no test specified\\\\\\" && exit 1\\",
    \\"db:up\\": \\"echo creating && npm run db:create && echo created && npm run db:migrate && echo migrated\\",
    \\"db:reset\\": \\"npm run db:drop && npm run db:up\\",
    \\"db:create\\": \\"sequelize db:create\\",
    \\"db:drop\\": \\"[[ $NODE_ENV == production ]] && exit 1 || sequelize db:drop\\",
    \\"db:migrate\\": \\"sequelize db:migrate\\",
    \\"db:rollback\\": \\"sequelize db:migrate:undo\\",
    \\"db:rollback:all\\": \\"sequelize db:migrate:undo:all\\",
    \\"build\\": \\"tsc\\",
    \\"start\\": \\"node ./dist/server.js\\",
    \\"dev\\": \\"tsc-watch --onSuccess \\\\\\"node ./dist/server.js\\\\\\"\\"
  },
  \\"author\\": \\"\\",
  \\"license\\": \\"ISC\\",
  \\"dependencies\\": {
    \\"pg\\": \\"^8.7.1\\",
    \\"pg-hstore\\": \\"^2.3.4\\",
    \\"sequelize\\": \\"^6.14.1\\"
  },
  \\"devDependencies\\": {
    \\"@types/node\\": \\"^16.11.7\\",
    \\"@types/validator\\": \\"^13.7.1\\",
    \\"sequelize-cli\\": \\"^6.4.1\\",
    \\"tsc-watch\\": \\"^4.6.0\\",
    \\"typescript\\": \\"^4.5.5\\"
  }
}

",
      "name": "package.json",
      "type": "FILE",
    },
    Object {
      "content": "# fields
This project was generated with [Sequelize UI](https://github.com/tomjschuster/sequelize-ui). The project is a simple [Node.js](https://nodejs.dev/) server with [Sequelize ORM](https://sequelize.org/).

Be sure to test all code for correctness and to test database migrations in a test environment before deploying to production.

## Running Project

### Prerequesites
- [Node.js](https://nodejs.dev/)
- [PostgreSQL](https://www.postgresql.org/)

### Setup
1. Install dependencies: \`npm install\`
2. Setup database: \`npm run db:up\`

### Run
- Local development: \`npm run dev\`
- Production build: \`npm run build && npm start\`

## Bug Reports
Please report any bugs with generated code at [Sequelize UI Issues](https://github.com/tomjschuster/sequelize-ui/issues).
",
      "name": "README.md",
      "type": "FILE",
    },
    Object {
      "content": "import http from 'http'
import db from './db'
import { initModels } from './models'

async function run() {
  initModels(db)
  const hostname = process.env.HOSTNAME || '127.0.0.1'
  const port = parseInt(process.env.PORT || '3000')
  
  const server = http.createServer((req, res) => {
    res.statusCode = 200
    res.setHeader('Content-Type', 'text/plain')
    res.end('Hello World')
  })
  
  server.listen(port, hostname, () => {
    console.log(\`Server running at http://\${hostname}:\${port}/\`)
  })
}

run()",
      "name": "server.ts",
      "type": "FILE",
    },
    Object {
      "content": "{
  \\"compilerOptions\\": {
    \\"target\\": \\"ES2020\\",
    \\"module\\": \\"commonjs\\",
    \\"lib\\": [\\"ES2020\\"],
    \\"outDir\\": \\"dist\\",
    \\"strict\\": true,
    \\"moduleResolution\\": \\"node\\",
    \\"esModuleInterop\\": true,
    \\"allowJs\\": true,
    \\"forceConsistentCasingInFileNames\\": true
  }
}
",
      "name": "tsconfig.json",
      "type": "FILE",
    },
  ],
  "name": "fields",
  "type": "DIRECTORY",
}
`;
